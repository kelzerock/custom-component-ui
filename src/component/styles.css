html,
body {
  height: 100%;
  margin: 0;
}

.wrapper {
  margin: 0 auto;
  max-width: 800px;
}
