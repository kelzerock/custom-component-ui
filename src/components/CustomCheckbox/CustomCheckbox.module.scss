.wrapper {
  display: inline-flex;
  align-items: center;
  padding: 2px 4px;

  &:hover,
  &:hover .checkbox {
    cursor: pointer;
  }

  &.disabled:hover,
  &.disabled:hover .checkbox {
    cursor: default;
  }
}


.checkbox {
  margin: 2px;
  accent-color: var(--color-main-blue);
}

.title {
  padding: 2px;
}