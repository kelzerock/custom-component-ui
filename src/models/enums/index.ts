export { Checkbox } from "./enums-for-test/Checkbox";
export { Switch } from "./enums-for-test/Switch";
export { Modal } from "./enums-for-test/Modal";
