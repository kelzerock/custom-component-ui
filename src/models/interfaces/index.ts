export { CustomSelectProps } from "./custom-select-props";
export { CustomButtonProps } from "./custom-button-props";
export { CustomTextFieldProps } from "./custom-text-field-props";
export { OptionsData } from "./option-data";
