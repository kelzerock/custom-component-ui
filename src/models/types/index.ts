export { ButtonSize } from "./button-size";
export { ButtonVariant } from "./button-variant";
