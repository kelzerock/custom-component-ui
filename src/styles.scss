:root {
  --color-main-blue: #1976d2;
  --color-main-blue-100: #2070c0;
  --color-white: #fff;
  --color-white-100: #f6fafd;
  --color-red: #d32f2f;
  --color-gray: #f0f0f0;
  --color-dark-gray: #00000099;
}