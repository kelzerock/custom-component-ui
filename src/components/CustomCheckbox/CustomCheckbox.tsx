import { FC, useState } from "react";
import { Checkbox } from "../../models/enums";
import { CustomCheckboxProps } from "../../models/interfaces";
import styles from "./CustomCheckbox.module.scss";
import clsx from "clsx";

export const CustomCheckbox: FC<CustomCheckboxProps> = ({
  title,
  color,
  onChange,
  checked,
  required,
  className,
  disabled,
  defaultChecked,
  ...props
}: CustomCheckboxProps) => {
  const requiredBlock = required ? " *" : "";
  const [checkedState, setCheckedState] = useState<boolean>(
    defaultChecked ?? false,
  );

  const handleClick = () => {
    if (disabled) return;

    setCheckedState((prev) => !prev);
  };

  return (
    <label
      className={clsx(styles.wrapper, disabled && styles.disabled)}
      data-testid={Checkbox.wrapper}
    >
      <input
        onChange={onChange ?? handleClick}
        type="checkbox"
        checked={checked ?? checkedState}
        className={clsx(styles.checkbox, className)}
        style={{ accentColor: color }}
        data-testid={Checkbox.checkbox}
        disabled={disabled}
        {...props}
      />
      {title && (
        <span
          className={styles.title}
          data-testid={Checkbox.title}
        >{`${title}${requiredBlock}`}</span>
      )}
    </label>
  );
};
