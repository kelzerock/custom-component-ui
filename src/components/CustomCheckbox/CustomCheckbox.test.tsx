import { render, screen } from "@testing-library/react";
import { CustomCheckbox } from "./CustomCheckbox";
import { Checkbox } from "../../models/enums";
import styles from "./CustomCheckbox.module.scss";
import userEvent from "@testing-library/user-event";

describe("check CustomCheckbox component", () => {
  it("All elements CustomCheckbox rendered", () => {
    const titleTest = "text";
    render(<CustomCheckbox title={titleTest} />);

    const wrapper = screen.getByTestId(Checkbox.wrapper);
    expect(wrapper).toBeInTheDocument();
    expect(wrapper).toHaveClass(styles.wrapper);

    const checkbox = screen.getByTestId(Checkbox.checkbox);
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).toHaveClass(styles.checkbox);

    const title = screen.getByTestId(Checkbox.title);
    expect(title).toBeInTheDocument();
    expect(title).toHaveClass(styles.title);
  });

  it("Title doesn't rendered without props", () => {
    render(<CustomCheckbox />);
    const title = screen.queryByTestId(Checkbox.title);
    expect(title).toBeNull();
  });

  it("Disable state correctly rendered", () => {
    render(<CustomCheckbox disabled />);
    const checkbox = screen.getByTestId(Checkbox.checkbox);
    expect(checkbox).toHaveAttribute("disabled");
  });

  it("Checked state correctly rendered when checked equal true", () => {
    render(<CustomCheckbox checked />);
    const checkbox = screen.getByTestId(Checkbox.checkbox);
    expect(checkbox).toHaveAttribute("checked");
  });

  it("Checked status correctly rendered when checked equal false", () => {
    render(<CustomCheckbox checked={false} />);
    const checkbox = screen.getByTestId(Checkbox.checkbox);
    expect(checkbox).not.toHaveAttribute("checked");
  });

  it("Correctly changed status checkbox", async () => {
    render(<CustomCheckbox />);
    const checkbox = screen.getByTestId(Checkbox.checkbox);
    expect(checkbox).toBeInTheDocument();

    if (!(checkbox instanceof HTMLInputElement)) {
      throw new Error("Checkbox is not an instance of HTMLInputElement");
    }

    expect(checkbox.checked).toBe(false);
    await userEvent.click(checkbox);
    expect(checkbox.checked).toBe(true);
    await userEvent.click(checkbox);
    expect(checkbox.checked).toBe(false);
  });

  it("toggles checkbox state when clicking on label", async () => {
    render(<CustomCheckbox title="Click me" />);
    const wrapper = screen.getByTestId(Checkbox.wrapper);
    const checkbox = screen.getByTestId(Checkbox.checkbox);

    if (!(checkbox instanceof HTMLInputElement)) {
      throw new Error("Checkbox is not an instance of HTMLInputElement");
    }

    expect(checkbox.checked).toBe(false);
    await userEvent.click(wrapper);
    expect(checkbox.checked).toBe(true);
  });

  test("check UI snapshot", () => {
    const { container } = render(<CustomCheckbox title="Test" color="green" />);
    expect(container).toMatchSnapshot();
  });
});
