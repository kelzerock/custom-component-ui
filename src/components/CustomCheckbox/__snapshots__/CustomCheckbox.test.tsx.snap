// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check CustomCheckbox component check UI snapshot 1`] = `
<div>
  <label
    class="wrapper"
    data-testid="wrapper"
  >
    <input
      class="checkbox"
      data-testid="checkbox"
      type="checkbox"
    />
    <span
      class="title"
      data-testid="title"
    >
      Test
    </span>
  </label>
</div>
`;
